src/http/httpClient/Middleware/BodyInspectionHandler.cs

```diff
@@ -0,0 +1,104 @@
+
+namespace Microsoft.Kiota.Http.HttpClientLibrary.Middleware;
+
+/// <summary>
+/// The Body Inspection Handler allows the developer to inspect the body of the request and response.
+/// </summary>
+public class BodyInspectionHandler : DelegatingHandler
+{
+    private readonly BodyInspectionHandlerOption _defaultOptions;
+
+    /// <summary>
+    /// Create a new instance of <see cref="BodyInspectionHandler"/>
+    /// </summary>
+    /// <param name="defaultOptions">Default options to apply to the handler</param>
+    public BodyInspectionHandler(BodyInspectionHandlerOption? defaultOptions = null)
+    {
+        _defaultOptions = defaultOptions ?? new BodyInspectionHandlerOption();
+    }
+
+    /// <inheritdoc/>
+    protected override async Task<HttpResponseMessage> SendAsync(
+        HttpRequestMessage request,
+        CancellationToken cancellationToken
+    )
+    {
+        if(request == null)
+            throw new ArgumentNullException(nameof(request));
+
+        var options = request.GetRequestOption<BodyInspectionHandlerOption>() ?? _defaultOptions;
+
+        Activity? activity;
+        if(request.GetRequestOption<ObservabilityOptions>() is { } obsOptions)
+        {
+            var activitySource = ActivitySourceRegistry.DefaultInstance.GetOrCreateActivitySource(
+                obsOptions.TracerInstrumentationName
+            );
+            activity = activitySource?.StartActivity(
+                $"{nameof(BodyInspectionHandler)}_{nameof(SendAsync)}"
+            );
+            activity?.SetTag("com.microsoft.kiota.handler.bodyInspection.enable", true);
+        }
+        else
+        {
+            activity = null;
+        }
+        try
+        {
+            if(options.InspectRequestBody)
+            {
+                options.RequestBody = await CopyToStreamAsync(request.Content, cancellationToken).ConfigureAwait(false);
+            }
+            var response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
+            if(options.InspectResponseBody)
+            {
+                options.ResponseBody = await CopyToStreamAsync(response.Content, cancellationToken).ConfigureAwait(false);
+            }
+
+            return response;
+        }
+        finally
+        {
+            activity?.Dispose();
+        }
+
+#if NETSTANDARD2_1_OR_GREATER || NETCOREAPP3_0_OR_GREATER
+        [return: NotNullIfNotNull(nameof(httpContent))]
+#endif
+        static async Task<Stream?> CopyToStreamAsync(HttpContent? httpContent, CancellationToken cancellationToken)
+        {
+            if(httpContent == null)
+            {
+                return null;
+            }
+
+            if(httpContent.Headers.ContentLength == 0)
+            {
+                return Stream.Null;
+            }
+
+            var stream = new MemoryStream();
+            await httpContent.CopyToAsync(stream, cancellationToken).ConfigureAwait(false);
```

Comment:
shouldn't we check whether the underlying streams are rewindable?
My concern here is the request body will be at its end position before it's sent over the network after the copy but before it's sent on the network.
Same thing the other way around with the response body and deserialization.

Reply:
When can that happen? How to check for such case?

Reply:
In unit tests, we're checking if  (after the copy) both request and response are still able to give us the proper content.

Reply:
I believe string contents are buffered by default, which would explain why this issue does not appear with the current tests.
Using a memory stream as the source and a different kind of content (application octet stream for example) might reveal it.
Can you try to expose the issue by tweaking the tests please?

Reply:
will do

Reply:
@baywet Have a look at https://github.com/microsoft/kiota-dotnet/pull/493/commits/5ed13f77b7c1cb94cb9ab794ce92a0085f74a588

Reply:
yes! thank you! resolving